codalab/server/bundle_manager.py

```diff
@@ -333,36 +336,42 @@ def _schedule_run_bundles_on_workers(self, workers, staged_bundles_to_run, user_
         # Dispatch bundles
         for bundle, bundle_resources in staged_bundles_to_run:
 
-            def get_available_workers(user_id):
-                # Make a deepcopy of the workers list so the filtering and deducting don't modify the list
-                workers_list = copy.deepcopy(workers.user_owned_workers(user_id))
-                workers_list = self._deduct_worker_resources(workers_list, running_bundles_info)
-                workers_list = self._filter_and_sort_workers(workers_list, bundle, bundle_resources)
+            def filter_workers_by_request_queue(user_id):
+                user_owned_workers = copy.deepcopy(workers.user_owned_workers(user_id))
+                if bundle.metadata.request_queue:
+                    workers_list = self._get_matched_workers(
+                        bundle.metadata.request_queue, user_owned_workers
+                    )
+                else:
+                    workers_list = user_owned_workers
                 return workers_list
 
-            workers_list = None
+            workers_list = []
             if bundle.owner_id != self._model.root_user_id:
                 # First try private workers
-                workers_list = get_available_workers(bundle.owner_id)
-                # If there is no user_owned worker, try to schedule the current bundle to run on a CodaLab's public worker.
-                if len(workers_list) == 0:
-                    # Check if there is enough parallel run quota left for this user
-                    if (
-                        self._model.get_user_parallel_run_quota_left(
-                            bundle.owner_id, user_info_cache[bundle.owner_id]
-                        )
-                        <= 0
-                    ):
-                        logger.info(
-                            "User %s has no parallel run quota left, skipping job for now",
-                            bundle.owner_id,
-                        )
-                        # Don't start this bundle yet, as there is no parallel_run_quota left for this user.
-                        continue
-            if not workers_list:
-                # Length is 0 (private user with no workers) or is None (root user)
-                workers_list = get_available_workers(self._model.root_user_id)
+                workers_list = filter_workers_by_request_queue(bundle.owner_id)
+
+            # If there is no user owned worker or the bundle is requested to run on CodaLab's public workers,
```

Comment:
Logic shouldn't make reference to a notion of public or not.

Reply:
Updated!